src/Constraint/MultiConstraint.php

```diff
@@ -82,23 +85,28 @@ public function isDisjunctive()
      */
     public function matches(ConstraintInterface $provider)
     {
+        $cacheKey = (string) $provider;
```

Comment:
Given that the result of parseConstraints is cached, a given constraint will only have one instance, so this could probably use spl_object_id as cacheKey instead of stringifying. Probably saves some CPU time generating __toString() in some cases, and possibly a little memory too.

Reply:
I thought about that, but this method is only available with php 7

Reply:
more over it would have work only if instance where the same (which may n'ot always be the case)

Reply:
Right, spl_object_hash would work on 5.2+, and well 99% of constraints in composer are created via parseConstraints so will be always the same instances. But that is correct strictly speaking it's pretty risky if we can't have a unique id per instance which will not be reused. That could be achieved by adding a static instance id counter and add a getUniqueId() or so which would be counter + classname, but not sure if it saves enough CPU to be worth it. Maybe try quickly with spl_object_hash if the saving seems worth the trouble of implementing anything more complex/safer?

Reply:
Hmm.. my concerns is also about cache miss (Constraint is not alway generated by parseConstraints)

For example, Composer calls the `matches` method with new Instance of Constraints for each package
https://github.com/composer/composer/blob/master/src/Composer/Repository/ComposerRepository.php#L234-L236

Reply:
be careful about `spl_object_id` and `spl_object_hash`. They are only a unique identifier among live objects. Once an object is destroyed, its id will be reused when instantiating new objects later. So using it as the cache key looks like a bad idea to me here, as you don't enforce keeping a reference to the constraint to keep it alive (which could easily leak memory).

Reply:
Indeed. Both risky and performance affecting (cache miss). `(string) $constraint` is still the best candidat :/ we have

Reply:
We should at least make sure that __toString caches it's output for constraints with no prettyString set then to avoid re-generating it every time.